buildingmotif/template.py

```diff
@@ -0,0 +1,219 @@
+        for param, value in bindings.items():
+            all_bindings[param] = value
+
+        # if all of the parameters have been bound then produce a serialized graph
+        # and then parse it to an rdflib.Graph
+        if len(bindings) == len(self.parameters):
+            g = new_graph(more_namespaces)
+            if more_namespaces:
+                preamble = PREAMBLE + "\n".join(
+                    f"@prefix {prefix}: <{uri}> .\n"
+                    for prefix, uri in more_namespaces.items()
+                )
+            else:
+                preamble = PREAMBLE
+            data = preamble + self.body.format(**all_bindings)
+            g.parse(data=data, format="ttl")
+            return g
+
+        # if *not* all of the parameters have been bound, then produce a new template
+        # with the provided elements filled in
+        partial = copy(self)
+        partial.body = self.body.format(**all_bindings)
+        partial.head = [x for x in self.head if x not in bindings.keys()]
+        return partial
+
+    def fill_in(self, bldg: Namespace) -> Graph:
+        """
+        Evaluates the template with autogenerated bindings w/n the given namespace
+        """
+        bindings = {param: f"bldg:{param}_{token_hex(8)}" for param in self.parameters}
+        res = self.evaluate(bindings, {"bldg": bldg})
+        assert isinstance(res, Graph)
+        return res
+
+    def __repr__(self):
+        return f"Template({self.name})"
+
+
+class TemplateLibrary:
+    def __init__(self, filename: Union[str, Path]) -> None:
+        self.templates = self._load_template_file(filename)
+
+    def __getitem__(self, key):
+        return self.templates[key]
+
+    def __repr__(self):
+        return f"TemplateLibrary({self.templates})"
+
+    def _load_template_file(self, filename) -> Dict[str, List[Template]]:
+        """
+        Returns list of Tempalte objets defined in the file
+        """
+        ret = defaultdict(list)
+        with open(filename, "r") as f:
+            data = yaml.load(f, yaml.Loader)
+            for templ in data:
+                for name, defn in templ.items():
+                    deps = {d["rule"]: d["args"] for d in defn.get("dependencies", [])}
+                    templ[name]["deps"] = deps
+                templ = Template(self, templ)
+                ret[templ.name].append(templ)
+        return ret
+
+
+def dump(templ, params, more_namespaces=None):
+    templ.inline_dependencies()
+    if params:
+        res = templ.evaluate(params, more_namespaces)
+    else:
+        res = templ.fill_in(Namespace("https://example.org/bldg"))
+    if isinstance(res, Graph):
+        print(res.serialize(format="ttl"))
+        print("ALL DONE!")
+    else:
+        print(
+            f"original params ({len(templ.parameters)}): {templ.parameters}"
+            "\nnow have ({len(res.parameters)}): {res.parameters}"
+        )
+
+
+if __name__ == "__main__":
```

Comment:
given the addition of tests, do we want to get rid of this?

Reply:
Good point; removed!